examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -41,101 +50,346 @@ def freq_mask(freqs, fmin, fmax):
+        # #sort in descencing order
+        ix = np.argsort(eigvals_)[::-1]
+        self.eigvals_=eigvals_[ix]
+        self.filters_ = eigvects_[:,ix]
+        # self.filters_ = eigvects_
+
+        self.patterns_ = np.linalg.pinv(self.filters_)      
+             
+                
+        return self
+ 
+    def spectral_ratio_ssd(self,ssd_sources):
+        """Spectral ratio measure for best n_components selection
+        See Nikulin 2011, Eq. (24)
+        
+        Parameters
+        ----------
+        ssd_sources : data projected on the SSD space. 
+        output of transform
+       
+        """
+                  
+            
+        psd, freqs = psd_array_welch(
+        ssd_sources, sfreq=self.sampling_freq, n_fft=self.n_fft)
+        sig_idx = _time_mask(freqs, *self.freqs_signal)
+        noise_idx = _time_mask(freqs, *self.freqs_noise)
+        if psd.ndim ==3:
+            spec_ratio = psd[:, :,sig_idx].mean(axis=2).mean(axis=0) / psd[:,:, noise_idx].mean(axis=2).mean(axis=0) 
 
-        self.eigvals_, self.filters_ = eigh(cov_signal.data, cov_noise.data)
-        self.patterns_ = np.linalg.pinv(self.filters_)
-
-    def transform(self, data):
-        if self.n_components is None:
-            n_components = len(self.picks_)
         else:
-            n_components = self.n_components
-        return np.dot(self.filters_.T[:n_components], data[self.picks_])
+                    
+            spec_ratio = psd[:, sig_idx].mean(axis=1) / psd[:, noise_idx].mean(axis=1)
+        sorter_spec = spec_ratio.argsort()[::-1]
+        return spec_ratio, sorter_spec
+
+    def transform(self, inst):
+        """Estimate epochs sources given the SSD filters.
+
+        Parameters
+        ----------
+        inst : instance of Raw or Epochs (n_epochs, n_channels, n_times)
+            The data to be processed. The instance is modified inplace.
+       
+        Returns
+        -------
+        out : instance of Raw or Epochs
+            The processed data.
+        
+        """
+        if self.filters_ is None:
+            raise RuntimeError('No filters available. Please first call fit')
+        if self.return_filtered:
+            inst=self.Xs
+        if isinstance(inst, BaseRaw):
+            data=inst.get_data()
+            X_ssd=np.dot(self.filters_.T, data[self.picks_])
+           
+        else:
+            if isinstance(inst, BaseEpochs)  or isinstance(inst, np.ndarray):
+                
+                data=inst    
+                #project data on source space
+                X_ssd = np.asarray([np.dot(self.filters_.T, epoch) for epoch in data])
+        
+        if self.sort_by_spectral_ratio:
+            self.spec_ratio, self.sorter_spec=self.spectral_ratio_ssd(ssd_sources=X_ssd)
+            self.filters_=self.filters_[:,self.sorter_spec]
+            self.patterns_=self.patterns_[self.sorter_spec]
+
+            if isinstance(inst, BaseRaw):
+                X_ssd=X_ssd[self.sorter_spec]
```

Comment:
white spaces .. all the way done. Please ensure the style is respecting whitespace rules elswhere.